xapian-glib/xapian-term-iterator.cc

```diff
@@ -0,0 +1,317 @@
+
+  iter = static_cast<XapianTermIterator *> (g_object_new (XAPIAN_TYPE_TERM_ITERATOR, NULL));
+  priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter);
+  priv->data = new IteratorData (it);
+
+  return iter;
+}
+
+/**
+ * xapian_term_iterator_is_valid:
+ * @iter: a #XapianTermIterator
+ *
+ * Checks whether @iter is valid.
+ *
+ * Returns: %TRUE if the iterator is valid
+ *
+ * Since: 1.4
+ */
+gboolean
+xapian_term_iterator_is_valid (XapianTermIterator *iter)
+{
+  g_return_val_if_fail (XAPIAN_IS_TERM_ITERATOR (iter), FALSE);
+
+  return XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter)->data != NULL;
+}
+
+/**
+ * xapian_term_iterator_next:
+ * @iter: a #XapianTermIterator
+ *
+ * Advances the @iter forward.
+ *
+ * You should use the return value to check if @iter is still
+ * valid after being advanced, before using the #XapianTermIterator
+ * API.
+ *
+ * Returns: %TRUE if the iterator was advanced, and %FALSE otherwise
+ *
+ * Since: 1.4
+ */
+gboolean
+xapian_term_iterator_next (XapianTermIterator *iter)
+{
+  g_return_val_if_fail (XAPIAN_IS_TERM_ITERATOR (iter), FALSE);
+
+  XapianTermIteratorPrivate *priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter);
+
+  if (priv->data == NULL)
+    return FALSE;
+
+  return priv->data->next ();
+}
+
+/**
+ * xapian_term_iterator_get_term_name:
+ * @iter: a #XapianTermIterator
+ *
+ * Retrieves the term name of the current item pointed by @iter.
+ *
+ * Returns: (transfer full): the term name
+ *
+ * Since: 1.4
+ */
+char *
+xapian_term_iterator_get_term_name (XapianTermIterator *iter)
+{
+  g_return_val_if_fail (XAPIAN_IS_TERM_ITERATOR (iter), 0);
+
+  XapianTermIteratorPrivate *priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter);
+
+  if (priv->data == NULL)
+    return NULL;
+
+  return priv->data->getTermName ();
+}
+
+/**
+ * xapian_term_iterator_get_wdf:
+ * @iter: a #XapianTermIterator
+ * @error: return location for a #GError
+ *
+ * Retrieves the WDF of the current item pointed by @iter.
+ *
+ * Returns: the WDF
+ *
+ * Since: 1.4
+ */
+unsigned int
+xapian_term_iterator_get_wdf (XapianTermIterator *iter,
+                              GError            **error)
+{
+  g_return_val_if_fail (XAPIAN_IS_TERM_ITERATOR (iter), 0);
+
+  XapianTermIteratorPrivate *priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter);
+
+  if (priv->data == NULL)
+    return 0;
+
+  try
+    {
+      return priv->data->getWdf ();
+    }
+  catch (const Xapian::Error &err)
+    {
+      GError *internal_error = NULL;
+
+      xapian_error_to_gerror (err, &internal_error);
+      g_propagate_error (error, internal_error);
+
+      return 0;
+    }
+}
+
+/**
+ * xapian_term_iterator_get_term_freq:
+ * @iter: a #XapianTermIterator
+ * @error: return location for a #GError
+ *
+ * Retrieves the term frequency of the current item pointed by @iter.
+ *
+ * Returns: the term frequency
+ *
+ * Since: 1.4
+ */
+unsigned int
+xapian_term_iterator_get_term_freq (XapianTermIterator *iter,
+                                    GError            **error)
```

Comment:
I did not mean `(unsigned int) -1`...
Leaving the `Xapian::doccount` inconsistency aside (I do think that it would be worth fixing while we can break ABI, but it sounds like something for another PR), my more immediate concern is that for methods that throw an error, there should be a way to detect that without actually passing a `GError` down; or in other words, if a `GError` is required to detect the error, the return value for an error should only be used in case of error.
A method that returns e.g. `0` both setting an error and not setting it is otherwise error-prone (pun intended!) to use.

In this particular case you can either avoid returning an error altogether, make sure that an error is always set when `0` is returned, or change the signature of the method so that the term frequency is returned as an out parameter, and the return value changed in a boolean (which would be `FALSE` when the error is set).